crypt.go

```diff
@@ -0,0 +1,494 @@
+
+	// RecoveryReasonPINFail indicates that a volume had to be activated with the fallback recovery key because the correct PIN was not
+	// provided.
+	RecoveryReasonPINFail
+)
+
+func activateWithRecoveryKey(volumeName, sourceDevicePath string, keyReader io.Reader, tries int, reason RecoveryReason, activateOptions []string) error {
+	if tries == 0 {
+		return errors.New("no recovery key tries permitted")
+	}
+
+	var lastErr error
+
+	for ; tries > 0; tries-- {
+		lastErr = nil
+
+		r := keyReader
+		keyReader = nil
+
+		passphrase, err := getPassword(sourceDevicePath, "recovery key", r)
+		if err != nil {
+			return xerrors.Errorf("cannot obtain recovery key: %w", err)
+		}
+
+		key, err := decodeRecoveryKey(passphrase)
+		if err != nil {
+			lastErr = xerrors.Errorf("cannot decode recovery key: %w", err)
+			continue
+		}
+
+		if err := activate(volumeName, sourceDevicePath, key, activateOptions); err != nil {
+			err = xerrors.Errorf("cannot activate volume: %w", err)
+			var e *exec.ExitError
+			if !xerrors.As(err, &e) {
+				return err
+			}
+			lastErr = err
+			continue
+		}
+
+		if _, err := unix.AddKey("user", fmt.Sprintf("%s:%s:reason=%d", filepath.Base(os.Args[0]), volumeName, reason), key, userKeyring); err != nil {
+			lastErr = xerrors.Errorf("cannot add recovery key to user keyring: %w", err)
+		}
+		break
+	}
+
+	return lastErr
+}
+
+func unsealKeyFromTPM(tpm *TPMConnection, k *SealedKeyObject, pin string) ([]byte, error) {
+	key, err := k.UnsealFromTPM(tpm, pin)
+	if err == ErrTPMProvisioning {
+		// ErrTPMProvisioning in this context might indicate that there isn't a valid persistent SRK. Have a go at creating one now and then
+		// retrying the unseal operation - if the previous SRK was evicted, the TPM owner hasn't changed and the storage hierarchy still
+		// has a null authorization value, then this will allow us to unseal the key without requiring any type of manual recovery. If the
+		// storage hierarchy has a non-null authorization value, ProvionTPM will fail. If the TPM owner has changed, ProvisionTPM might
+		// succeed, but UnsealFromTPM will fail with InvalidKeyFileError when retried.
+		if pErr := ProvisionTPM(tpm, ProvisionModeWithoutLockout, nil); pErr == nil {
+			key, err = k.UnsealFromTPM(tpm, pin)
+		}
+	}
+	return key, err
+}
+
+var requiresPinErr = errors.New("no PIN tries permitted when a PIN is required")
+
+type lockAccessError struct {
+	err error
+}
+
+func (e lockAccessError) Error() string {
+	return e.err.Error()
+}
+
+func (e lockAccessError) Unwrap() error {
+	return e.err
+}
+
+func isLockAccessError(err error) bool {
+	var e lockAccessError
+	return xerrors.As(err, &e)
+}
+
+func activateWithTPMKey(tpm *TPMConnection, volumeName, sourceDevicePath, keyPath string, pinReader io.Reader, pinTries int, lock bool, activateOptions []string) error {
+	var lockErr error
+	key, err := func() ([]byte, error) {
+		defer func() {
+			if !lock {
+				return
+			}
+			lockErr = LockAccessToSealedKeys(tpm)
```

Comment:
this code path doesn't seem tested

Reply:
I'm not sure how to test this yet - it's difficult to hit any of the error conditions in a test.